crates/revm/revm-inspectors/src/tracing/mod.rs

```diff
@@ -0,0 +1,408 @@
+            pc,
+            op: OpCode::try_from_u8(interp.contract.bytecode.bytecode()[pc])
+                .expect("is valid opcode;"),
+            contract: interp.contract.address,
+            stack: interp.stack.clone(),
+            memory: interp.memory.clone(),
+            gas: self.gas_inspector.as_ref().gas_remaining(),
+            gas_refund_counter: interp.gas.refunded() as u64,
+
+            // fields will be populated end of call
+            gas_cost: 0,
+            state_diff: None,
+            status: InstructionResult::Continue,
+        });
+    }
+
+    /// Fills the current trace with the output of a step.
+    ///
+    /// Invoked on [Inspector::step_end].
+    fn fill_step_on_step_end<DB: Database>(
+        &mut self,
+        interp: &mut Interpreter,
+        data: &mut EVMData<'_, DB>,
+        status: InstructionResult,
+    ) {
+        let StackStep { trace_idx, step_idx } =
+            self.step_stack.pop().expect("can't fill step without starting a step first");
+        let step = &mut self.traces.arena[trace_idx].trace.steps[step_idx];
+
+        if let Some(pc) = interp.program_counter().checked_sub(1) {
+            let op = interp.contract.bytecode.bytecode()[pc];
+
+            let journal_entry = data
+                .journaled_state
+                .journal
+                .last()
+                // This should always work because revm initializes it as `vec![vec![]]`
+                // See [JournaledState::new](revm::JournaledState)
+                .expect("exists; initialized with vec")
+                .last();
+
+            step.state_diff = match (op, journal_entry) {
+                (
+                    opcode::SLOAD | opcode::SSTORE,
+                    Some(JournalEntry::StorageChange { address, key, .. }),
+                ) => {
+                    // SAFETY: (Address,key) exists if part if StorageChange
+                    let value = data.journaled_state.state[address].storage[key].present_value();
+                    Some((*key, value))
+                }
+                _ => None,
+            };
+
+            step.gas_cost = step.gas - self.gas_inspector.as_ref().gas_remaining();
+        }
+
+        // set the status
+        step.status = status;
+    }
+}
+
+impl<DB> Inspector<DB> for TracingInspector
+where
+    DB: Database,
+{
+    fn initialize_interp(
+        &mut self,
+        interp: &mut Interpreter,
+        data: &mut EVMData<'_, DB>,
+        is_static: bool,
+    ) -> InstructionResult {
+        self.gas_inspector.initialize_interp(interp, data, is_static)
+    }
+
+    fn step(
+        &mut self,
+        interp: &mut Interpreter,
+        data: &mut EVMData<'_, DB>,
+        is_static: bool,
+    ) -> InstructionResult {
+        if self.record_steps {
+            self.gas_inspector.step(interp, data, is_static);
+            self.start_step(interp, data);
+        }
+
+        InstructionResult::Continue
+    }
```

Comment:
Hooking on step is really expensive I think even for just having an `if` branch when running a large number of transactions when `record_steps = false`..we should explore converting this into a const that we don't pay for at runtime eventually, similar to how Revm has a "const INSPECT: bool" for the entire thing